src/provider/provider_fixed.c

```diff
@@ -0,0 +1,714 @@
+        assert(alloc->data);
+        assert(alloc->size > 0);
+
+        // only the HEAD could have an empty prev
+        if (alloc != provider->alloc_list) {
+            assert(alloc->prev);
+        }
+
+        // check double-linking
+        if (alloc->prev) {
+            assert(alloc->prev->next == alloc);
+        }
+
+        if (alloc->next) {
+            assert(alloc->next->prev == alloc);
+        }
+
+        // data addresses in the list are in ascending order
+        if (alloc->prev) {
+            assert(alloc->prev->data < alloc->data);
+        }
+
+        if (alloc->next) {
+            assert(alloc->data < alloc->next->data);
+        }
+
+        // data should not overlap
+        if (alloc->next) {
+            assert((alloc->data + alloc->size) <= alloc->next->data);
+        }
+
+        sum_allocs_size += alloc->size;
+
+        alloc = alloc->next;
+    }
+
+    assert(sum_allocs_size == provider->alloc_size);
+}
+
+static enum umf_result_t fixed_memory_provider_initialize(void *params,
+                                                          void **provider) {
+    void *init_buffer = NULL;
+
+    if (provider == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (params == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    fixed_memory_provider_params_t *fixed_params =
+        (fixed_memory_provider_params_t *)params;
+
+    fixed_memory_provider_t *fixed_provider =
+        (fixed_memory_provider_t *)malloc(sizeof(fixed_memory_provider_t));
+
+    if (!fixed_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    // check params
+    if (fixed_params->upstream_memory_provider == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (pthread_mutex_init(&fixed_provider->lock, NULL) != 0) {
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    fixed_provider->trace = fixed_params->trace;
+    fixed_provider->upstream_memory_provider =
+        fixed_params->upstream_memory_provider;
+    fixed_provider->debug_check = fixed_params->debug_check;
+    fixed_provider->block_list = NULL;
+    fixed_provider->alloc_list = NULL;
+    fixed_provider->alloc_size = 0;
+    fixed_provider->used_size = 0;
+
+    if (fixed_params->immediate_init) {
+        fixed_memory_provider_alloc(
+            fixed_provider, fixed_params->init_buffer_size, 0, &init_buffer);
+
+        if (init_buffer == NULL) {
+            pthread_mutex_destroy(&fixed_provider->lock);
+            free(fixed_provider);
+            return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+        }
+
+        // since we use alloc func, we have set the block as unused
+        fixed_provider->used_size = 0;
+        fixed_provider->block_list->used = false;
+        fixed_provider->alloc_size = fixed_params->init_buffer_size;
+    }
+
+    *provider = fixed_provider;
+
+    // TODO check results
+    debug_check(fixed_provider);
```

Comment:
I would suggest implementing the `debug_check` without any asserts (just return true if everything is consistent, false otherwise) and call `assert(debug_check(fixed_provider))`. This way, you'll be sure no code will be executed in the Release build.

Reply:
I leave asserts but I call this function only inside the assert too, as you suggested